lib/ukstore/store.c

```diff
@@ -29,3 +29,1267 @@
+	case UK_STORE_ENTRY_TYPE(u8): {
+		__u8 val;
+
+		str = calloc(_U8_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.u8(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _U8_STRLEN, "%" __PRIu8, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(s8): {
+		__s8 val;
+
+		str = calloc(_S8_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.s8(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _S8_STRLEN, "%" __PRIs8, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(u16): {
+		__u16 val;
+
+		str = calloc(_U16_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.u16(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _U16_STRLEN, "%" __PRIu16, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(s16): {
+		__s16 val;
+
+		str = calloc(_S16_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.s16(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _S16_STRLEN, "%" __PRIs16, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(u32): {
+		__u32 val;
+
+		str = calloc(_U32_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.u32(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _U32_STRLEN, "%" __PRIu32, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(s32): {
+		__s32 val;
+
+		str = calloc(_S32_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.s32(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _S32_STRLEN, "%" __PRIs32, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(u64): {
+		__u64 val;
+
+		str = calloc(_U64_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.u64(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _U64_STRLEN, "%" __PRIu64, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(s64): {
+		__s64 val;
+
+		str = calloc(_S64_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.s64(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _S64_STRLEN, "%" __PRIs64, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(uptr): {
+		__uptr val;
+
+		str = calloc(_UPTR_STRLEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.uptr(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		snprintf(str, _UPTR_STRLEN, "0x%" __PRIx64, val);
+		*out  = str;
+		return ret;
+	}
+
+	case UK_STORE_ENTRY_TYPE(charp): {
+		char *val = NULL;
+
+		ret = e->get.charp(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		*out  = val;
+		return ret;
+	}
+
+	default:
+		return -EINVAL;
+	}
+}
+
+
+int
+_uk_store_get_ncharp(const struct uk_store_entry *e, char **out, size_t maxlen)
+{
+	int ret;
+
+	UK_ASSERT(e);
+
```

Comment:
In all the getter variabts and setter variants you should make sure that a getter is available. Otherwise return `-ENOENT`. Since get is a union you can do a non-NULL check of just one preferred field. Please add this to all getter and setter API functions.

Reply:
I can just check if the enum is `_LIB_UK_STORE_ENTRY_TYPE___undef`  -  I will do it like this

Reply:
```
	if (unlikely(e->type == UK_STORE_ENTRY_TYPE(undef)))
		return -ENOENT;
```

Reply:
But doesn't it already go in the default if the type is something other then the ones used?

```
	default:
		return -EINVAL;
	}
```

Reply:
It is for the case that we have get-only or set-only entries. Actually I would return -EIO if you set on a get-only entry (and vice versa). I think EIO is more appropriate for this case